keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,819 @@
+	// It can return an error if the field selector cannot be parsed.
+    // Remember that field selector formats vary based on the version of the API being used!
+	ParseFieldSelector() ([]Requirement, error)
+  
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+    // It can return an error if the field selector cannot be parsed.
+    ParseLabelSelector() ([]Requirement, error)
+```
+
+#### Future-proofing your authentication webhook for future verbs
+
+As of 1.31, the only verbs with field and label selectors are List, Watch, and DeleteCollection.
+<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>
+In the future, the kube-apiserver may add field and label selectors to Create, Update, and Delete.
+<<[/UNRESOLVED]>>
+* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 
+* For Update, this means that the finalNewObject and oldObject must match the field and label selector.
+* For Delete, this means that the oldObject must match the field and label selector.
+ 
+We do not expand field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`
+field selector and use a List to get equivalent functionality.
+
+### SubjectAccessReview Changes
+
+SubjectAccessReview is used for two purposes:
+1. Authorization webhook calls from the kube-apiserver to a webhook.
+   This usage likely benefits from a serialization with `[]Requirement`.
+2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)
+   This usage likely benefits from a serialization that matches the query parameter.
+
+Their needs best met with two different serialization (see user stories)
+
+```go
+type SubjectAccessReviewSpec struct {
+	FieldSelector *FieldSelectorAttributes
+	
+	LabelSelector *LabelSelectorAttributes
+}
+
+// FieldSelectorAttributes indicates a field limited access.
+// For webhooks:
+// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the request is not limited.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
+// For the kube-apiserver:
+// If rawSelector is empty and requirements are empty, the request is not limited.
+// If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
+// If rawSelector is empty and requirements are present, the requirements should be honored
+// If rawSelector is present and requirements are present, the request is invalid.
```

Comment:
could these be simplified to:

```
* If `rawSelector` and `requirements` are both present, the request is invalid.
* For kube-apiserver: if `rawSelector` is present, the rawSelector will be parsed and honored if the parsing succeeds and the verb supports selectors.
* For kube-apiserver and webhooks: if requirements are present, the requirements should be honored if the verb supports selectors.
```

Reply:
> the requirements should be honored if the verb supports selectors.

How and why would an authorizer make a decision about whether the client that called it supports selectors for a particular verb?  If we make this a webhook responsibility, would this mean that the webhook needs to predict when the kube-apiserver will honor selectors for verb and subresource tuples?

Reply:
I find the fully separation much easier to find what I'm looking for.  Your feel strongly about simplifying?  Do you want the change the to webhook semantics from what I proposed?  I don't see how it's possible.